repository/src/main/java/org/alfresco/repo/security/authentication/identityservice/SpringBasedIdentityServiceFacade.java

```diff
@@ -1,344 +1,352 @@
+                AuthorizationGrantType.PASSWORD, createPasswordClient(restOperations, clientRegistration));
+    }
+
+    @Override
+    public AccessTokenAuthorization authorize(AuthorizationGrant authorizationGrant)
+    {
+        final AbstractOAuth2AuthorizationGrantRequest request = createRequest(authorizationGrant);
+        final OAuth2AccessTokenResponseClient client = getClient(request);
+
+        final OAuth2AccessTokenResponse response;
+        try
+        {
+            response = client.getTokenResponse(request);
+        }
+        catch (OAuth2AuthorizationException e)
+        {
+            LOGGER.debug("Failed to authorize against Authorization Server. Reason: " + e.getError() + ".");
+            throw new AuthorizationException("Failed to obtain access token. " + e.getError(), e);
+        }
+        catch (RuntimeException e)
+        {
+            LOGGER.warn("Failed to authorize against Authorization Server. Reason: " + e.getMessage());
+            throw new AuthorizationException("Failed to obtain access token.", e);
+        }
+
+        return new SpringAccessTokenAuthorization(response);
+    }
+
+    @Override
+    public Optional<OIDCUserInfo> getUserInfo(String tokenParameter, String principalAttribute)
+    {
+        return Optional.ofNullable(tokenParameter)
+                .filter(Predicate.not(String::isEmpty))
+                .flatMap(token -> Optional.ofNullable(clientRegistration)
+                        .map(ClientRegistration::getProviderDetails)
+                        .map(ClientRegistration.ProviderDetails::getUserInfoEndpoint)
+                        .map(ClientRegistration.ProviderDetails.UserInfoEndpoint::getUri)
+                        .flatMap(uri -> {
+                            try
+                            {
+                                return Optional.of(
+                                        new UserInfoRequest(new URI(uri), new BearerAccessToken(token)).toHTTPRequest().send());
+                            }
+                            catch (IOException | URISyntaxException e)
+                            {
+                                LOGGER.warn("Failed to get user information. Reason: " + e.getMessage());
+                                return Optional.empty();
+                            }
+                        })
+                        .flatMap(httpResponse -> {
+                            try
+                            {
+                                UserInfoResponse userInfoResponse = UserInfoResponse.parse(httpResponse);
+
+                                if (userInfoResponse instanceof UserInfoErrorResponse)
+                                {
+                                    UserInfoErrorResponse userInfoErrorResponse = (UserInfoErrorResponse) userInfoResponse;
+                                    String errorMessage = userInfoErrorResponse.getErrorObject().getDescription();
```

Comment:
In general, it looks good to me.
Please replace `UserInfoErrorResponse userInfoErrorResponse = (UserInfoErrorResponse) userInfoResponse;`
with pattern variable:
`if (userInfoResponse instanceof UserInfoErrorResponse userInfoErrorResponse)`.
Moreover, I think we can check if `userInfoErrorResponse.getErrorObject()` is `null`.
For example: 
```
String errorMessage = Optional.ofNullable(userInfoErrorResponse.getErrorObject())
                                            .map(ErrorObject::getDescription)
                                            .orElse("No error description found");
```

Reply:
I have addressed the following changes